proposals/type-applications-in-patterns.rst

```diff
@@ -0,0 +1,170 @@
+
+
+A more complex example is this (also inspired by _#15050 <https://ghc.haskell.org/trac/ghc/ticket/15050>_)::
+
+    data T a where
+      MkT1 :: forall a.              T a
+      MkT2 :: forall a.              T (a,a)
+      MkT3 :: forall a b. b ~ Int => T a
+      MkT4 :: forall a b.            T a
+      MkT5 :: forall a b c. b ~ c => T a
+      
+    foo :: T (Int, Int) -> ()
+    foo (MkT1 @(Int,Int))  = ()
+    foo (MkT2 @x)          = (() :: x ~ Int => ())
+    foo (MkT3 @_ @Int)     = ()
+    foo (MkT4 @_ @x)       = (() :: x ~ x => ()) -- (these constraints here just to
+    foo (MkT5 @_ @x @x)    = (() :: x ~ x => ()) --  demonstrate that x is in scope)
+
+All of these equations type-check (just like they would if added value arguments of type ``a``, ``b``,... to the constructors and turned the type applications into type signatures).
+
+This example demonstrated why we need to typecheck nested patterns left-to-right::
+
+ data T a where
+   T1 :: T Int
+   T2 :: T a
+
+ f :: Int -> Char -> Bool
+
+ g :: (a, Char, T a) -> blah
+ g (x :: Int, (f x -> True), T1) = ..
+
+``g`` must not be accepted: Until we match on ``T1`` we have no idea if ``a ~ Int``.
+And, with Haskell's left-to-right pattern matching we'll
+match the view pattern ``(f x -> True)`` first. It looks ok, because
+you can see that ``x :: Int``; but it will seg-fault in a call of
+``g`` involving ``T2`` and a first argument that is (say) a list.
+
+Scoping
+~~~~~~~
+
+The scoping works just like with ``ScopedTypeVariables``. Just for reference, here are some examples of how that feature works now::
+
+ f :: forall a b. ([a], b) -> INt
+ f (x :: [v], y) = ...
+
+brings ``v`` into scope, together with ``a`` and ``b``, which are already in scope.
+
+But the pattern in::
+
+ f :: forall a b. ([a], b) -> INt
+ f (x :: [b], y) = ...
+
+does not bring ``b`` into scope; here ``b`` refers to the ``b`` from the type signature.
+
+And the pattern in::
+ 
+ f :: forall a b. ([a], b) -> INt
+ f (x :: [v], y :: v) = ...
+
+brings one ``v`` into scope; the second occurence in the pattern does not shadow the first one, but rather refers to the same type (this would lead to a type error because ``v`` needs to be equal to both ``a`` and ``b``, but maybe they are not the same).
+
+The same rules apply for type applications, and similarly to the last example, the following should not type-check:
+
+ data T where
+   MkT :: a -> b -> T
+
+ f (MkT @p @p a b) = ...
+
+Effect and Interactions
+-----------------------
+By motivation our answer the question of “what should ``@ty`` mean in patterns” with an existing feature (type signatures in patterns), we fill an obvious hole in the syntax in a way that is consistent with existing features: The analogy between type applications and type signatures will hold the same way in terms as it would in types.
+
+Furthermore, type application arguments to ``C`` refer to the corresponding parameters in both terms and types (which
```

Comment:
I think I know what you mean here, but I'm not sure.   Perhaps point to the alternative you have in mind?  And/or a section of the paper?

Reply:
Rewritten.